src/graphnet/data/sqlite_dataconverter.py

```diff
@@ -151,17 +151,17 @@ def _parallel_extraction(self, settings):
                 # Concatenate data
                 for key, data in data_dict.items():
                     df = apply_event_no(data, event_no_list, event_count)
-                    if len(df):
+                    if bool( data_dict[self._pulsemap]['dom_x'] and data_dict[self._table_names[0]] and len(df) ): #only include if the dom_x is non-empty and the truth is non-empty
```

Comment:
1. I think a somewhat detailed comment could be useful here, to explain the logic behind this statement.
2. This seems to assume that the truth extractor is at index 0, but that may not always be the case.
3. Also: Do we _want_ to require the existence of truth data for writing pulsemaps in data (where truth doesn't exist, unless I'm mistaken)?
4. I don't think the outer `bool(...)` statement is needed, as there are three implicit conversions to bools happening. If anything, I think it could be more useful to make the three boolean conversion explicit instead (e.g., `len(df) > 0`).
5. See also comment above: What happens in the case where we have multiple pulsemaps in the same event (e.g., Upgrade) where you may have hits in one but not the other, which, if you're unlucky, could be the first one.

Reply:
Agree almost totally with this comment. I would just say:
1.Definitely, I'll start writing such a description in a subsequent commit.
2.The position of the truth extractor is actually asserted in L60 but I am not opposed to this changing, if understandability improves.
3. There is a check in the I3TruthExtractor is_mc, that guards against trying to extract truth from real data and leaves a table in databases filled with padding values (but the all important event_no's). As things currently stand we don't allow databases to be made without a truth extractor, this could be changed but serious work would be required to ensure attached indices were unique (particularly if data is split across multiple databases)
4. Can, and will, change. I thought it was worthwhile making it explicit that this is a empty check. But I think `len(df) > 0` is even more clear.
5. A very good point, one that I didn't anticipate.  This method still supports multiple pulse maps being extracted but only if the first non-truth non-retro key is non-empty which I guess could be called the "master" pulsemap. If we take out the pulsemap check (i.e just check the truth and current key) then this leads to undesirable outcomes such as events with just truth and retro (for example) tables, which happens a lot in level 2 data since many P frames are "NullSplit" and contain truth, retro (etc.) and no pulse maps.